src/modules/bootloader/main.py

```diff
@@ -265,19 +290,30 @@ def install_systemd_boot(efi_directory):
     subprocess.call(["bootctl",
                      "--path={!s}".format(install_efi_directory),
                      "install"])
-    create_systemd_boot_conf(install_path,
-                             efi_directory,
-                             uuid,
-                             distribution,
-                             distribution_translated,
-                             "default")
-    if "fallback" in libcalamares.job.configuration:
-        create_systemd_boot_conf(install_path,
-                                 efi_directory,
-                                 uuid,
-                                 distribution,
-                                 distribution_translated,
-                                 "fallback")
+
+    # find all the installed kernels and generate default and fallback entries for each
+    # This is Arch-specific and may need adjustment for other distros
+    for root, dirs, files in os.walk(os.path.join(install_path, kernel_search_path.lstrip('/'))): 
+        for file in files:  
+            if file == source_kernel_name:
+                rel_root = os.path.relpath(root, install_path)
```

Comment:
The four lines above are used to generate two variables for use in the calls to `create_systemd_boot_conf()`, namely *file* and *rel_root*.  Then there's two variants / flavors, namely "default" and "fallback".

A function returning a list of 3-tuples could factor out some of the Arch-ness: on arch, do this `os.walk()`, elsewhere return a list with just the default flavor (and maybe "fallback").

Reply:
That is a great idea.  If we can devise a strategy for resolving the mkinitcpio problem above, there would be nothing left that isn't generic.